apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/HiPerfPublisher.cs

```diff
@@ -509,98 +513,148 @@ private void ShutdownIfCompleted()
+                            batch.BatchCompletion.SetCanceled();
                         }
-                        else
-                        {
-                            _shutdownTcs.SetResult(0);
-                        }
-                    });
+                    }
                 }
+                // All batches sent and shutdown requested, so signal shutdown completed successfully.
+                _taskHelper.Run(async () =>
+                {
+                    if (_shutdown != null)
+                    {
+                        await _taskHelper.ConfigureAwaitHideErrors(_shutdown());
+                    }
+                    if (_hardStopCts.IsCancellationRequested)
+                    {
+                        _shutdownTcs.SetCanceled();
+                    }
+                    else
+                    {
+                        _shutdownTcs.SetResult(0);
+                    }
+                });
             }
         }
 
-        // Not "async void" to gain standard async exeption handling
-        private async Task CreateBatch()
+        private void DelaySendCurrentBatch()
         {
-            // Pre-condition: Must be locked - lock active up until first 'await'.
-            _currentBatch = new Batch();
-            var timerToken = _currentBatch.TimerCts.Token;
-            // After time-delay threshold, send the batch regardless of fullness.
-            await _taskHelper.ConfigureAwait(_scheduler.Delay(_batchDelayThreshold, timerToken));
-            // If batch has already moved to queue, timerToken will have been cancelled.
-            lock (_lock)
+            // Pre-condition: Must be locked
+            if (_batchDelayThreshold is TimeSpan batchDelayThreshold)
             {
-                // Check for cancellation inside lock to avoid race-condition.
-                if (!timerToken.IsCancellationRequested)
+                // read cancellation token here, in case the current batch changes before the task below starts.
+                var timerCancellationToken =
+                    CancellationTokenSource.CreateLinkedTokenSource(_currentBatch.TimerCts.Token, _hardStopCts.Token).Token;
+                // Ignore result of this Task. If it's cancelled, it's because the batch has already been sent.
+                _taskHelper.Run(async () =>
                 {
-                    // Force queuing of the current batch, whatever the size.
-                    // There will always be at least one message in the batch.
-                    QueueCurrentBatchIfRequired(true, 0);
-                }
+                    await _taskHelper.ConfigureAwait(_scheduler.Delay(batchDelayThreshold, timerCancellationToken));
+                    // If batch has already moved to queue, timerToken will have been cancelled.
+                    lock (_lock)
+                    {
+                        // Check for cancellation inside lock to avoid race-condition.
```

Comment:
It may be worth considering making the batch know whether or not it's been queued, and just unconditionally call some "QueueIfNotAlreadyQueued" here - and rather than using _currentBatch, either read _currentBatch before the task, or accept it as a method parameter.

No action required for now, but an alternative to consider for later.